### PR TITLE
V2/bug/ffn input dim

### LINKED ISSUE #481: [BUG]: readout input size is wrong in `train.py` for additional molecular features
**Describe the bug**
Per Angiras and Xiarui:
> Xiaorui and I are having a look at additional features and are encountering a dimension error when it comes to readout when using the Morgan fingerprints. The output dimension of the message passing block becomes 2348 (default 300 + 2048 from Morgan fingerprint) but the readout input dimension is still 300. In train.py it seems that the readout input dimension is set to be mp.block.output_dim which seems to correspond to the size before any additional features are concatenated. As a result, it seems that the readout input dimension is unaware of the additional features vector x_f at the moment. In v1, there was another arg use_input_features that would then be added to the first linear dimension during readout and was set when additional features through features_generators or features_path were passed. Do we want to adopt the same approach for v2, or was there another approach you wanted to take?

### PR DESCRIPTION
**Pull request by davidegraff**
## Description
This PR fixes a bug in FFN construction by improving some of the dataset internals

## Example / Current workflow
Currently, this block
https://github.com/chemprop/chemprop/blob/9650df0ceedf6c04f32e1ad54d54bb179deebca1/chemprop/v2/cli/train.py#L802-L812
produces a downstream error during modelling because it doesn't account for possible additional features that will be passed during property readout. That is, if you have additional molecular features of size `d_xf` you will need the input to your readout network to be of size `mp_block.output_dim + d_xf`, which is clearly not happening here.

## Bugfix / Desired workflow
What this PR does to address is primarily introduce a new dataset attribute `d_xf` which is equal to the size of the additional feature dimension. In the case there are no additional features, this will be `0`. There are corresponding attributes: `d_ef`, `d_vf`, and `d_vd` that already exist and returned `None` in similar cases. To be logically consistent, these have been changed to `0` as well. This is logically correct because if we have no additional {edge,node} features, then the size of these features is `0`. While it might be tempting to say `None` because these features/descriptors don't exist, I argue that this leads to an edge case where we have added empty "features" of size 0 and our code will treat this differently. Abstractly, these two are the same thing and should be treated accordingly.

One small thing this PR also does is change is during batch collation it now checks the entire batch of additional features/descriptors being `None` rather than just the 0-th element. This is functionally the same, but could undercover corner cases where some of these items are set in the batch and some are `None`. Our code is not currently equipped to handle that, and it'll error out earlier and more clearly.

## Relevant issues
closes #481 

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added/fixed?

### CODE REVIEW

**Comment by kevingreenman** [comment]
lgtm. We can merge if @xiaoruiDong and/or @am2145 agree

File: chemprop/v2/data/datasets.py [hunk 1]
```diff
@@ -80,6 +81,11 @@ def lt_mask(self) -> np.ndarray:
     def t(self) -> int | None:
         return self.data[0].t if len(self.data) > 0 else None
 
+    @property
+    def d_xf(self) -> int:
+        """the extra molecule feature dimension, if any"""
+        return 0 if np.equal(self.X_f, None).all() else self.V_fs[0].shape[1]
```
Comment (xiaoruiDong) [inline chemprop/v2/data/datasets.py:87]: `self.X_f.shape[1]` instead of `self.V_fs[0].shape[1]`?

File: chemprop/v2/cli/train.py [hunk 2]
```diff
@@ -801,7 +802,7 @@ def main(args):
 
     readout_ffn = Factory.build(
         readout_cls,
-        input_dim=mp_block.output_dim,
```
Comment (xiaoruiDong) [inline chemprop/v2/cli/train.py:804]: I guess this is a separate issue.

I ran a test and found that the `input_dim` is still `300` despite the change here. I found that the factory method does not work as expected for `RegressionFFN` (while it works for `BinaryClassificationFFN`). I think the issue is due to the fact that `RegressionFFN` has `__init__` redefined, and different from `ReadoutFFNBase`. Then, in `build`:

https://github.com/chemprop/chemprop/blob/e777044bfc46bfb14927a53bc89b366dfa17acab/chemprop/v2/utils/registry.py#L39-L46

the obtained `sig.parameters.keys()` is actually `odict_keys(['args', 'loc', 'scale', 'kwargs'])` that doesn't include `input_dim` and other arguments...

**Review by xiaoruiDong — CHANGES_REQUESTED** [review]
Thank you, @davidegraff, for implementing these. Please take a look at my comments. Besides, I think you also need to change the following:

https://github.com/chemprop/chemprop/blob/e777044bfc46bfb14927a53bc89b366dfa17acab/chemprop/v2/models/model.py#L77-L81

As with additional molecular features `X_f`, it will be`message_passing.out_dim + X_f.shape == readout.input_dim`. However, the dimension info of `X_f` is not given to `MPNN`...

File: chemprop/v2/cli/train.py (hunk 2 above)
Comment (davidegraff) [reply chemprop/v2/cli/train.py:804]: good catch! I've changed all `ReadoutFFN` (sub)classes to now use explicit `__init__` signatures to fix this.

**Comment by davidegraff** [comment]
re: the comment on the dimensionality agreement for `MPNN.__init__`, I've deleted this check. In line with other general torch modules, we now just assume that the shapes work for a client's use-case, and if it fails, well then that's their problem. An analogy would be doing the following:
```python
net = nn.Sequential([nn.Linear(128, 64), nn.ReLU(), nn.Linear(128, 1)])
```
PyTorch will let you construct this erroneous module even though it could reasonably check for dimensionality agreement here. We can still check for this in the CLI, but we shouldn't check inside the class itself.

File: chemprop/v2/tests/unit/featurizers/test_atom.py
Comment (am2145) [inline chemprop/v2/tests/unit/featurizers/test_atom.py:1]: The atom features array for `test_x_orig` has been spread to have one element per line. Could you undo this?

Comment (davidegraff) [reply chemprop/v2/tests/unit/featurizers/test_atom.py:1]: fixed now

File: chemprop/v2/data/datasets.py
Comment (am2145) [inline chemprop/v2/data/datasets.py:1]: Line 88: `self.X_f[0].shape[1]` should be `self.X_f.shape[1]`?

**Comment by davidegraff** [comment]
@xiaoruiDong @am2145 seems like changes have been addressed?

**Review by xiaoruiDong — APPROVED** [review]
LGTM. Thanks!

**Comment by am2145** [comment]
Everything looks good to me. Thanks David!
